### PR TITLE
Security Info Popup closes #454

### LINKED ISSUE #454: Show information about SSL certificate
To fix this bug implement the mocks that @brampitoyo provides. See https://github.com/mozilla-mobile/focus-android/issues/454#issuecomment-304553461 and later comments.

Please note this is slated for v5.0 and will need to be completed by mid-Februrary, according to [the current calendar](https://wiki.mozilla.org/Mobile/Focus/Android/Train_Schedule).

---

From feedback:

> Seems like I cannot view SSL cert info. Pressing the lock icon in Fennec Android shows some info.

cert

### PR DESCRIPTION
**Pull request by ekager**

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/popup/PopupFactory.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+import org.mozilla.focus.R;
+import org.mozilla.focus.session.Session;
+import org.mozilla.focus.session.SessionManager;
+import org.mozilla.focus.utils.DrawableUtils;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class PopupFactory {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupFactory.java]: \[PMD\] INFO: The class 'PopupFactory' has a Cyclomatic Complexity of 11 \(Highest = 10\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupFactory.java]: \[PMD\] INFO: The class 'PopupFactory' has a Modified Cyclomatic Complexity of 11 \(Highest = 10\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupFactory.java]: \[PMD\] INFO: The class 'PopupFactory' has a Standard Cyclomatic Complexity of 11 \(Highest = 10\)\.

File: app/src/main/java/org/mozilla/focus/popup/PopupFactory.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    public static PopupWindow createSecurityPopup(Context context) {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupFactory.java]: \[PMD\] INFO: The method 'createSecurityPopup' has a Cyclomatic Complexity of 10\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupFactory.java]: \[PMD\] INFO: The method 'createSecurityPopup' has a Modified Cyclomatic Complexity of 10\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupFactory.java]: \[PMD\] INFO: The method 'createSecurityPopup' has a Standard Cyclomatic Complexity of 10\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupFactory.java]: \[PMD\] INFO: The method createSecurityPopup\(\) has an NPath complexity of 222

File: app/src/main/java/org/mozilla/focus/popup/PopupFactory.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        final Session session = SessionManager.getInstance().getCurrentSession();
+        if (session != null) {
+            if (session.getSecure().getValue() != null) {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupFactory.java]: \[PMD\] INFO: These nested if statements could be combined

File: app/src/main/java/org/mozilla/focus/popup/SecurityPopup.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+import android.content.Context;
+import android.content.res.Resources;
+import android.graphics.Color;
+import android.graphics.drawable.Drawable;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.webkit.URLUtil;
+import android.widget.FrameLayout;
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/SecurityPopup.java]: \[Checkstyle\] ERROR: Unused import \- android\.widget\.FrameLayout\.

File: app/src/main/java/org/mozilla/focus/popup/SecurityPopup.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import org.mozilla.focus.session.Session;
+import org.mozilla.focus.utils.DrawableUtils;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class SecurityPopup {
+    public static PopupWindow createSecurityPopup(final Context context, final Session session) {
+        if (session != null) {
+            if (session.getSecure().getValue() != null) {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/SecurityPopup.java]: \[PMD\] INFO: These nested if statements could be combined

File: app/src/main/java/org/mozilla/focus/popup/PopupUtils.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+import android.view.ViewGroup;
+import android.webkit.URLUtil;
+import android.widget.ImageView;
+import android.widget.PopupWindow;
+import android.widget.TextView;
+
+import org.mozilla.focus.R;
+import org.mozilla.focus.session.Session;
+import org.mozilla.focus.utils.DrawableUtils;
+
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The class 'PopupUtils' has a Cyclomatic Complexity of 11 \(Highest = 10\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The class 'PopupUtils' has a Modified Cyclomatic Complexity of 11 \(Highest = 10\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The class 'PopupUtils' has a Standard Cyclomatic Complexity of 11 \(Highest = 10\)\.

File: app/src/main/java/org/mozilla/focus/popup/PopupUtils.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import java.net.MalformedURLException;
+import java.net.URL;
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The method 'createSecurityPopup' has a Cyclomatic Complexity of 10\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The method 'createSecurityPopup' has a Modified Cyclomatic Complexity of 10\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The method 'createSecurityPopup' has a Standard Cyclomatic Complexity of 10\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The method createSecurityPopup\(\) has an NPath complexity of 222

File: app/src/main/java/org/mozilla/focus/popup/PopupUtils.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+public class PopupUtils {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The class 'PopupUtils' has a Cyclomatic Complexity of 11 \(Highest = 10\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The class 'PopupUtils' has a Modified Cyclomatic Complexity of 11 \(Highest = 10\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The class 'PopupUtils' has a Standard Cyclomatic Complexity of 11 \(Highest = 10\)\.

File: app/src/main/java/org/mozilla/focus/popup/PopupUtils.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    public static PopupWindow createSecurityPopup(final Context context, final Session session) {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The method 'createSecurityPopup' has a Cyclomatic Complexity of 10\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The method 'createSecurityPopup' has a Modified Cyclomatic Complexity of 10\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The method 'createSecurityPopup' has a Standard Cyclomatic Complexity of 10\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: \[PMD\] INFO: The method createSecurityPopup\(\) has an NPath complexity of 222

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 10]
```diff
@@ -285,13 +292,24 @@ public void onValueChanged(@NonNull Boolean loading) {
...
-                lockView.setVisibility(secure ? View.VISIBLE : View.GONE);
+                if (secure) {
+                    lockView.setImageResource(R.drawable.ic_lock);
+                } else {
+                    if (URLUtil.isHttpUrl(getUrl())) {
+                        // HTTP site
+                        lockView.setImageResource(R.drawable.ic_internet);
+                    } else {
+                        // Certificate is bad
+                        lockView.setImageResource(R.drawable.ic_warning_16);
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: Let's remove the "_16" suffix from the image files. For a vector drawable that doesn't mean much.

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 11]
```diff
@@ -1045,6 +1069,25 @@ public void setBlockingEnabled(boolean enabled) {
...
+        final PopupWindow securityPopup = PopupUtils.createSecurityPopup(getContext(), session);
+        if (securityPopup != null) {
+            securityPopup.setOnDismissListener(new PopupWindow.OnDismissListener() {
+                @Override
+                public void onDismiss() {
+                    popupTint.setVisibility(View.GONE);
+                }
+            });
+            securityPopup.setAnimationStyle(android.R.style.Animation_Dialog);
+            popupTint.setVisibility(View.VISIBLE);
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: over-nitpicking. You have a bunch of calls on "securityPopup" here and in between a call on popupTint. To make this more readable I'd move this line to the end.

File: app/src/main/java/org/mozilla/focus/popup/PopupUtils.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+import android.widget.TextView;
+
+import org.mozilla.focus.R;
+import org.mozilla.focus.session.Session;
+import org.mozilla.focus.utils.DrawableUtils;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class PopupUtils {
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: No Kotlin? 😭

File: app/src/main/java/org/mozilla/focus/popup/PopupUtils.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+    public static PopupWindow createSecurityPopup(final Context context, final Session session) {
+        if (session.getSecure().getValue() != null) {
+            final boolean isSecure = session.getSecure().getValue();
+            final String url = session.getUrl().getValue();
+            final Resources res = context.getResources();
+
+            final LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/popup/PopupUtils.java]: nit: There's also a shortcut: LayoutInflater.from(context).

File: app/src/main/res/drawable/ic_check_16.xml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+<vector android:height="16dp" android:viewportHeight="16.0"
```
Comment (pocmo) [inline app/src/main/res/drawable/ic_check_16.xml]: nit: license header (a bunch of our existing XML files are missing them too)

File: app/src/main/res/layout/security_popup.xml [hunk 15]
```diff
@@ -0,0 +1,66 @@
+<?xml version="1.0" encoding="utf-8"?><!-- This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (pocmo) [inline app/src/main/res/layout/security_popup.xml]: nit: line break before the license header. Android Studio messes this up when using the auto-formatter. :)

File: app/src/main/res/layout/security_popup.xml [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+       - License, v. 2.0. If a copy of the MPL was not distributed with this
+       - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (pocmo) [inline app/src/main/res/layout/security_popup.xml]: Is the wrapping RelativeLayout needed? If yes, could it be a FrameLayout instead (performance)?

File: app/src/main/res/values/strings.xml [hunk 17]
```diff
@@ -555,4 +555,9 @@ be temporary, and you can try again later.</li>
     <string name="gv_prompt_username_hint">Username</string>
     <string name="gv_prompt_password_hint">Password</string>
     <string name="gv_prompt_clear">Clear</string>
+
+    <string name="secure_connection">Secure Connection</string>
+    <string name="insecure_connection">Insecure Connection</string>
+    <string name="security_verified">Verified by: %1$s</string>
+    <string name="site_security">Site Security</string>
```
Comment (pocmo) [inline app/src/main/res/values/strings.xml]: Can you add some comments here explaining what the Strings are used for - or where they show up? Comments *before* a String definition will show up in the tools our translators use and help them to come up with a good translation.

nit: I wonder if we should use a prefix here, just to group the strings, e.g. security_popup_*

**Comment by pocmo** [comment]
From testing:

* Something doesn't feel right about the icon sizes. A quick look at the files: The new ones seem to use width/height:16dp viewportWidth/Height:16.0 - is that causing this?

* While a (secure) page is still loading I see the "warning" icon. Only after the page has finished loading the "Secure" icon is shown. What is causing this? Could we keep showing the globe until we know for sure?

**Comment by pocmo** [comment]
The complexity complains from sputnik are kind of valid. createSecurityPopup() is pretty long and does a lot of things. Maybe we could split it up into some smaller private methods with nice names? :)

**Comment by ekager** [comment]
Pulled svgs from a different location and seems to have fixed those issues!

Changed it show be the internet icon while still loading.

**Comment by pocmo** [comment]
Landed separately to fix one conflict and two complaints from detekt: cece775f0cce6abea733fbf9a73b86d10e253816

**Pull request closed without merging**